File: week-02/009-april_26/product_except_self.cpp
/* ---------------------------------------
 *                   DAY 9
 *         Apple Interview Question
 *
 *            PRODUCT EXCEPT SELF
 *  Given an array nums of n integers where
 *  n > 1, return an array such that out[i]
 *  is equal to the product of all elements
 *  except nums[i].
 *
 *  by: @estalvgs1999
 * ---------------------------------------- */

#include <iostream>
#include <vector>
using namespace std;

vector<int> product_except_self(vector<int> nums){

    int n = nums.size();
    
    vector<int> out(n);

    out[0] = 1;

    for(int i=1; i < n; i++){
        out[i] = nums[i-1]*out[i-1];
    }

    int R = 1;
    for(int i=n-1; i>=0; i--){
        out[i] = out[i]*R;
        R = R*nums[i];
    }

    return out;
}

int main(int argc, char const *argv[])
{
    vector<int> nums = {1,2,3,4};
    vector<int> out = product_except_self(nums);

    for (vector<int>::iterator it = out.begin() ; it != out.end(); ++it)
        cout << *it<<" "; 
    cout<<endl;
    return 0;
}

File: week-02/009-april_26/product_except_self.py
# ---------------------------------------
#                   DAY 9
#         Apple Interview Question
#
#            PRODUCT EXCEPT SELF
#  Given an array nums of n integers where
#  n > 1, return an array such that out[i]
#  is equal to the product of all elements
#  except nums[i].
#
#  by: @estalvgs1999
# ----------------------------------------

# Brute Force Solution
# Complexity O(n²)
def bf_product_except_self(nums):

    output = []

    for i in range(len(nums)):
        current_product = 1
        for j in range(len(nums)):
            if i != j:   
                current_product *= nums[j]
        output.insert(i,current_product)

    return output

# Using division Solution 
# Complexity O(2n)
def ud_product_except_self(nums):

    p = 1
    for n in nums: p*=n
    output = [p//n for n in nums]
    
    return output

# Solution without division
# Complexity O(2n)
def product_except_self(nums):

    n = len(nums)
    output = [1]*n

    for i in range(1,n):
        output[i] = nums[i-1]*output[i-1]

    R = 1
    for i in range(n-1,-1,-1):
        output[i] = output[i]*R
        R = R*nums[i]

    return output



if __name__ == "__main__":
    print(bf_product_except_self([1,2,3,4]))
    print(ud_product_except_self([1,2,3,4]))
    print(product_except_self([1,2,3,4]))